File: cli/src/fp_cli/commands/pull.py
"""fp pull — pull app code to local directory."""

from pathlib import Path

import typer

from ..api.client import PlatformClient, PlatformError, resolve_app
from ..config import get_active_platform_or_exit
from ..console import console, err_console
from ..project import write_project, PROJECT_FILE


def pull(
    name: str = typer.Argument(..., help="App name to pull"),
    force: bool = typer.Option(False, "--force", "-f", help="Overwrite existing files"),
):
    """Pull app code from the platform to the current directory."""
    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    app = resolve_app(client, name)

    try:
        detail = client.get_app(app["app_id"])
    except PlatformError as e:
        err_console.print(f"[red]Failed to get app details:[/red] {e.message}")
        raise typer.Exit(1)

    cwd = Path.cwd()

    # Check for existing .fp.yaml
    if (cwd / PROJECT_FILE).exists() and not force:
        err_console.print(
            f"[yellow]{PROJECT_FILE} already exists.[/yellow] Use [bold]--force[/bold] to overwrite."
        )
        raise typer.Exit(1)

    mode = detail.get("mode", "single")
    written = 0

    if mode == "multi" and detail.get("files"):
        files = detail["files"]
        # Prefer deployed files, fall back to draft
        if detail.get("deployed_files"):
            files = detail["deployed_files"]
        for filename, content in files.items():
            filepath = cwd / filename
            filepath.parent.mkdir(parents=True, exist_ok=True)
            filepath.write_text(content)
            written += 1
    elif detail.get("code") or detail.get("deployed_code"):
        code = detail.get("deployed_code") or detail.get("code", "")
        (cwd / "app.py").write_text(code)
        written = 1
    else:
        err_console.print("[red]App has no code to pull.[/red]")
        raise typer.Exit(1)

    # Write project file
    entrypoint = detail.get("entrypoint") or "app.py"
    project_data = {"name": detail["name"], "entrypoint": entrypoint}
    write_project(project_data)

    console.print(f"Pulled [bold]{detail['name']}[/bold] ({written} file{'s' if written != 1 else ''})")
    console.print(f"  Written to: {cwd}")

File: cli-go/cmd/pull.go
package cmd

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/api"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/project"
)

var pullCmd = &cobra.Command{
	Use:   "pull [name]",
	Short: "Pull app code from the platform to the current directory",
	Args:  cobra.ExactArgs(1),
	RunE:  runPull,
}

var pullForce bool

func init() {
	pullCmd.Flags().BoolVarP(&pullForce, "force", "f", false, "Overwrite existing files")
}

func runPull(cmd *cobra.Command, args []string) error {
	appName := args[0]
	p := requirePlatform()
	client := api.NewClient(p.URL, p.Token)
	app, err := api.ResolveApp(client, appName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	detail, err := client.GetApp(app["app_id"].(string))
	if err != nil {
		var pe *api.PlatformError
		if errors.As(err, &pe) {
			fmt.Fprintf(os.Stderr, "Failed to get app details: %s\n", pe.Message)
			os.Exit(1)
		}
		return err
	}

	cwd, _ := os.Getwd()
	if _, err := os.Stat(filepath.Join(cwd, project.ProjectFile)); err == nil && !pullForce {
		fmt.Fprintf(os.Stderr, "%s already exists. Use --force to overwrite.\n", project.ProjectFile)
		os.Exit(1)
	}

	mode, _ := detail["mode"].(string)
	if mode == "" {
		mode = "single"
	}
	written := 0

	if mode == "multi" {
		files, _ := detail["deployed_files"].(map[string]interface{})
		if files == nil {
			files, _ = detail["files"].(map[string]interface{})
		}
		if files != nil {
			for filename, content := range files {
				if s, ok := content.(string); ok {
					fpath := filepath.Join(cwd, filename)
					if err := os.MkdirAll(filepath.Dir(fpath), 0755); err != nil {
						return err
					}
					if err := os.WriteFile(fpath, []byte(s), 0644); err != nil {
						return err
					}
					written++
				}
			}
		}
	} else {
		code, _ := detail["deployed_code"].(string)
		if code == "" {
			code, _ = detail["code"].(string)
		}
		if code != "" {
			if err := os.WriteFile(filepath.Join(cwd, "app.py"), []byte(code), 0644); err != nil {
				return err
			}
			written = 1
		}
	}

	if written == 0 {
		fmt.Fprintln(os.Stderr, "App has no code to pull.")
		os.Exit(1)
	}

	entrypoint, _ := detail["entrypoint"].(string)
	if entrypoint == "" {
		entrypoint = "app.py"
	}
	name, _ := detail["name"].(string)
	if err := project.Write(cwd, &project.Project{Name: name, Entrypoint: entrypoint}); err != nil {
		return err
	}

	plural := "s"
	if written == 1 {
		plural = ""
	}
	fmt.Printf("Pulled %s (%d file%s)\n", name, written, plural)
	fmt.Printf("  Written to: %s\n", cwd)
	return nil
}
